learn/models/conformal-regression/index.qmd

```diff
@@ -0,0 +1,441 @@
+The results for this method are:
+
+```{r}
+#| label: cv-pred
+#| out-width: 70%
+#| echo: false
+#| fig-width: 6
+#| fig-height: 4
+
+test_cv_res %>% 
+  ggplot(aes(x)) +
+  geom_point(aes(y = y), alpha = 1 / 20) +
+  geom_line(
+    data = nnet_pred, aes(y = .pred),
+    linewidth = 1, col = "blue"
+  ) +
+  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
+              col = "#D95F02", linewidth = 3 / 4, fill = NA)
+```
+
+At each point, the interval length is `r round(mean(test_cv_res$.pred_upper - test_cv_res$.pred_lower), 2)`; the previous split conformal interval width was `r round(mean(test_split_res$.pred_upper - test_split_res$.pred_lower), 2)`. This is due to the training set being larger than the calibration set. 
+
+Note that the coverage is a little better too: 
+
+
+```{r}
+#| label: cv-coverage
+coverage(test_cv_res)
+```
+
+## Adaptive width intervals
+
+The simulated data that we've been using has a constant error variance. That has worked for the two methods shown since their intervals are always the same width. Real data does not always have constant variation. 
+
+To demonstrate, we can take the previous simulation system and induce a variance that is dynamic over the predictor range: 
+
+```{r}
+#| label: variable-plot
+#| out-width: 70%
+#| fig-width: 6
+#| fig-height: 4
+
+make_variable_data <- function(n, std_dev = 1 / 5) {
+  tibble(x = runif(n, min = -1)) %>%
+    mutate(
+      y = (x^3) + 2 * exp(-6 * (x - 0.3)^2),
+      y = y + rnorm(n, sd = std_dev * abs(x))
+    )
+}
+
+make_variable_data(1000) %>% 
+  ggplot(aes(x, y)) + 
+  geom_point(alpha = 1 / 5)
+```
+
+Let's create new data sets and re-fit our model: 
+
+```{r}
+#| label: variable-data
+set.seed(7292)
+train_variable_data  <- make_variable_data(n)
+cal_variable_data  <- make_variable_data(250)
+test_variable_data <- make_variable_data(10000)
+
+nnet_variable_fit <- nnet_wflow %>% fit(train_variable_data)
+nnet_variable_pred <- augment(nnet_variable_fit, train_variable_data)
+```
+
+We can recreate the CV+ interval for this new version of the data:
+
+```{r}
+#| label: cv-redux
+#| out-width: 70%
+#| echo: false
+#| fig-width: 6
+#| fig-height: 4 
+set.seed(493)
+folds_variable <- vfold_cv(train_variable_data)
+
+nnet_variable_rs <- 
+  nnet_wflow %>% 
+  fit_resamples(folds_variable, control = ctrl)
+
+cv_variable_int <- int_conformal_cv(nnet_variable_rs)
+
+test_cv_variable_res <- 
+  predict(cv_variable_int, test_variable_data, level = 0.90) %>% 
+  bind_cols(test_variable_data)
+
+test_cv_variable_res %>% 
+  ggplot(aes(x)) +
+  geom_point(aes(y = y), alpha = 1 / 20) +
+  geom_line(
+    data = nnet_variable_pred, aes(y = .pred),
+    linewidth = 1, col = "blue"
+  ) +
+  geom_ribbon(aes(ymin = .pred_lower, ymax = .pred_upper), 
+              col = "#D95F02", linewidth = 1, fill = NA)
+```
+
+While the coverage is good (`r round(coverage(test_cv_variable_res)$coverage, 1)`%), there are obvious areas where it is either too wide or too narrow. 
```

Comment:
Is there a reason why you stopped running `coverage(test_cv_variable_res)` for people to see?